Implement heuristic-based split reduction value selection for `topk` op

grypp opened the pull request:
The current implementation of the `topk` operation lacks parallelism, causing performance issues on massively parallel architectures like GPUs.

The proposed change sets the split reduction values heuristically for the topk operation. The heuristic function analyzes the workload and balances the work between the initial topk kernel and the subsequent split topk kernel.

For example, when the input is a tensor of size 4x8000 and the value of K is 40, current implementation the last dimension (tensor<8000>) is run sequentially. The heuristic function selects 16 and 5, the amount of sequential work is significantly reduced, as demonstrated below:
```
topk2d_dispatch_0_topk_4x16x500xf32
topk2d_dispatch_1_generic --> this is entirely parallel, heuristic function doesn't check 
topk2d_dispatch_2_topk_4x5x128xf32
topk2d_dispatch_3_topk_4x200xf32
```

It is important to note that split reduction operates at the flow level, and the heuristic function is integrated there. While technically the heuristic function can split more depths, we limit it to two depths as the target architecture at the flow level is unknown.

By the way, the flag `iree-flow-topk-split-reduction` remains available. A user can opt-out the heuristic and set any value.

Note that, this is an interim solution and will not give peak performance. To achieve peak, it is crucial to have a GPU-optimized topk operation.

qcolombet commented on compiler/src/iree/compiler/Dialect/Flow/Transforms/SplitReduction.cpp:
```diff
@@ -76,14 +82,87 @@ struct LinalgSplitReduction
+                                            topkSplitReductionRatio.end());
+    if (splitReductionDepth >= reductionRatios.size()) {
+      return kNOTopkSplitReductionRatioDefault;
+    } else {
```
I know this part is just a refactoring, but no else after return.

qcolombet commented on compiler/src/iree/compiler/Dialect/Flow/Transforms/SplitReduction.cpp:
```diff
@@ -76,14 +82,87 @@ struct LinalgSplitReduction
+
+  // Step 1. Hard to predict of advantage of splitting reduction more than 2
+  // depth.
+  if (splitReductionDepth > 1) return kNOTopkSplitReductionRatioDefault;
```
Should we use the `reductionRatios` here instead of the default?

I.e., I believe that's more in line with what the code used to do. (Could be wrong though)

grypp replied:
`reductionRatios` is filled by the `topkSplitReductionRatio` flag. 

I think we really want to return here `kNOTopkSplitReductionRatioDefault` (that is `-1`) to not do further split reduction.

qcolombet replied:
Sounds good.

qcolombet commented on compiler/src/iree/compiler/Dialect/Flow/Transforms/SplitReduction.cpp:
```diff
@@ -76,14 +82,87 @@ struct LinalgSplitReduction
+    for (int64_t ki = 3; ki < kValue; ki++) {
+      if (workload % ki != 0) continue;
+      int workload_2nd_kernel = kValue * ki;
+      int workload_1st_kernel = workload / ki;
```
Nit: I think the IREE style is `camelCase`, i.e., no `_`

qcolombet commented on compiler/src/iree/compiler/Dialect/Flow/Transforms/SplitReduction.cpp:
```diff
@@ -76,14 +82,87 @@ struct LinalgSplitReduction
+                     << "Check Kernel = " << check << "\n";
+      });
+      if (workload_2nd_kernel > workload_1st_kernel &&
+          workload_2nd_kernel > check) {
```
Could you add a comment explaining the logic why this is a good split?
Also do you believe we can keep the logic target independent like this?
For instance, do you think we should have one heuristic for the GPU and one for the CPU?

grypp replied:
The heuristic attempts to split work as much as evenly between the dispatch regions. It splits with maximum depth of 2. I think it is okay for CPU and GPU.

CPU has single dimensional parallelism (excluding numa), so splitting with 2 depth would be fine. 

GPU has two dimensional parallelism and each dimension has 3 more dimensions (blockIdx x,y,z), so splitting with 2 depth is not the best choice. 

I don't know the target device at this phase. Therefore, I kept the split depth at 2.

qcolombet replied:
Okay, that's what I expected: not all targets would like the same splits.
Couple of follow-up questions:
- @grypp if we were to tune that heuristic solely for NV GPU, how much more perf could we get?
- @MaheshRavishankar what is the recommended way to plug this kind of target specific tweaks?

MaheshRavishankar replied:
Just coming online... The layering of these has been off from the get go. So this should be moved into plugin/pre-processing for now to make quick progress. The more structural issue will require more time (but happy to help).

grypp replied:
If tune this specifically in nvgpu, I could add additional parallelism, but the benefit would be negligible. This was the case from the cases I tried.

qcolombet replied:
@jpienaar how do you think we should approach these target specific heuristics?
For now, we can go crazy in the plugins, but flagging this one in case this sparks some ideas on how we should do things in the platform.

jpienaar replied:
Moving to plugin or pre-processing pass for now is POR.

More target specific heuristics should be queryable and here it should just be an invocation of a returned function pointer. The main consideration is 1) what attributes are queryable, 2) interface of predictor, 3) where. Fixing that and then we will do something a la const-eval. I'll write up a one pager here later, but the above in short.

grypp committed:
Implement heuristic-based split reduction value selection for topk op

The current implementation of the `topk` operation lacks parallelism, causing performance issues on massively parallel architectures like GPUs.

The proposed change sets the split reduction values heuristically for the topk operation. The heuristic function analyzes the workload and balances the work between the initial topk kernel and the subsequent split topk kernel.

For example, when the input is a tensor of size 4x8000 and the value of K is 40, current implementation the last dimension (tensor<8000>) is run sequentially. The heuristic function selects 16 and 5, the amount of sequential work is significantly reduced, as demonstrated below:
```
topk2d_dispatch_0_topk_4x16x500xf32
topk2d_dispatch_1_generic --> this is entirely parallel, heuristic function doesn't check
topk2d_dispatch_2_topk_4x5x128xf32
topk2d_dispatch_3_topk_4x200xf32
```

It is important to note that split reduction operates at the flow level, and the heuristic function is integrated there. While technically the heuristic function can split more depths, we limit it to two depths as the target architecture at the flow level is unknown.

By the way, the flag `iree-flow-topk-split-reduction` remains available. A user can opt-out the heuristic and set any value.

Note that, this is an interim solution and will not give peak performance. To achieve peak, it is crucial to have a GPU-optimized topk operation.

grypp committed:
address comments

MaheshRavishankar reviewed (changes requested):
Reviewing, but this needs to be guarded by a flag in its current usage... I think this is being made on by default here for all backends. So marking as request change while I look more.

MaheshRavishankar reviewed (dismissed):
There are a few things to consider here... 
1) This pass was setup to be a no-op when the `splitRatio` was -1. So that was the way this pass was "turned on/off", by setting the split value. IIUC this PR is moving away from that. We probably dont want to enable this unconditionally for all backends without actually making sure this works for all backends. So to make progress on the pass as is, could you control this pass using a command line flag here https://github.com/openxla/iree/blob/5efe2d763038f6af637384e11983ef5c6fc2fe7f/compiler/src/iree/compiler/Dialect/Flow/Transforms/Passes.cpp#L255 . This basically will allow you to adjust the heuristic if you need to without accidentally turning this on by default.

2) We need to move this pass into a plugin/pre-processing if this is to be turned on by default for NVGPU backends. This same pass is used for both splitting "k" of reduction and top-k. So this needs some care..

We can do (1) before to unblock and make progress before we get a good story for (2) in place.

qcolombet reviewed (approved):
Let’s go with this heuristic as the default

grypp committed:
Reduce depth to 1

grypp committed:
revert depth to 2

jpienaar commented on compiler/src/iree/compiler/Dialect/Flow/Transforms/SplitReduction.cpp:
```diff
@@ -76,14 +82,86 @@ struct LinalgSplitReduction
+/// Find there is a topk operation in the given operation.
+static bool hasTopk(Operation *op) {
+  bool hasTopk = false;
+  op->walk([&](LinalgExt::TopkOp topkOp) { hasTopk = true; });
```
Should this interrupt when it finds one rather than continue?

grypp committed:
interrupt at hasTopk

grypp committed:
add a test
